### PR TITLE
Add some element.*Attribute functions

### PR DESCRIPTION
**Pull request by krichprollsch**
Relates to https://github.com/Browsercore/browsercore/issues/20

### CODE REVIEW

File: src/dom/attribute.zig [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+const Node = @import("node.zig").Node;
+const DOMException = @import("exceptions.zig").DOMException;
+
+// WEB IDL https://dom.spec.whatwg.org/#attr
+pub const Attr = struct {
+    pub const Self = parser.Attribute;
+    pub const prototype = *Node;
+    pub const mem_guarantied = true;
+
+    pub const Exception = DOMException;
```
Comment (francisbouvier) [inline src/dom/attribute.zig]: It's not mandatory to repeat the `Exception` here as `Attr` is a child of `EventTarget` who has it already.

File: src/dom/element.zig [hunk 2]
```diff
@@ -2,16 +2,24 @@ const std = @import("std");
...
 
+const DOMException = @import("exceptions.zig").DOMException;
+
 // WEB IDL https://dom.spec.whatwg.org/#element
 pub const Element = struct {
     pub const Self = parser.Element;
     pub const prototype = *Node;
     pub const mem_guarantied = true;
 
+    pub const Exception = DOMException;
```
Comment (francisbouvier) [inline src/dom/element.zig]: Same here, we do not need the `Exception` declaration.

File: src/dom/namednodemap.zig [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+const checkCases = jsruntime.test_utils.checkCases;
+
+const DOMException = @import("exceptions.zig").DOMException;
+
+// WEB IDL https://dom.spec.whatwg.org/#namednodemap
+pub const NamedNodeMap = struct {
+    pub const Self = parser.NamedNodeMap;
+    pub const mem_guarantied = true;
+
+    pub const Exception = DOMException;
```
Comment (francisbouvier) [inline src/dom/namednodemap.zig:16]: Here we need it as `NamedNodeMap` is not a child of `Event Target`.

File: src/netsurf.zig [hunk 4]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+}
+
+pub fn namedNodeMapItem(nnm: *NamedNodeMap, index: u32) !?*Attribute {
+    var n: [*c]c.dom_node = undefined;
+    const err = c._dom_namednodemap_item(nnm, index, &n);
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]: I think we could inline those kind of checks, as this a common pattern and the file is already quite verbose.

File: src/netsurf.zig [hunk 5]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+}
+
+pub fn namedNodeMapGetNamedItem(nnm: *NamedNodeMap, qname: []const u8) !?*Attribute {
+    var n: [*c]c.dom_node = undefined;
+    const err = c._dom_namednodemap_get_named_item(nnm, try stringFromData(qname), &n);
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]: Inliner

File: src/netsurf.zig [hunk 6]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+        nnm,
+        try stringFromData(namespace),
+        try stringFromData(localname),
+        &n,
+    );
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]: Inliner

File: src/netsurf.zig [hunk 7]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+
+pub fn namedNodeMapGetLength(nnm: *NamedNodeMap) !u32 {
+    var ln: u32 = undefined;
+    const err = c.dom_namednodemap_get_length(nnm, &ln);
+    try DOMErr(err);
+    return ln;
+}
+
+pub fn namedNodeMapItem(nnm: *NamedNodeMap, index: u32) !?*Attribute {
+    var n: [*c]c.dom_node = undefined;
```
Comment (francisbouvier) [inline src/netsurf.zig:350]: We already talked about creating an internal alias `NodeInternal` to avoid using c pointer. Let's do that in a separate PR.

File: src/netsurf.zig [hunk 8]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+    const err = c._dom_namednodemap_item(nnm, index, &n);
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
+
+    // cast [*c]c.dom_node into *Attribute
```
Comment (francisbouvier) [inline src/netsurf.zig:356]: I think once we have the `NodeInternal` type we can remove those comments who are not so useful.

File: src/netsurf.zig [hunk 9]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+
+pub fn namedNodeMapGetNamedItemNS(
+    nnm: *NamedNodeMap,
+    namespace: []const u8,
+    localname: []const u8,
+) !?*Attribute {
+    var n: [*c]c.dom_node = undefined;
+    const err = c._dom_namednodemap_get_named_item_ns(
+        nnm,
+        try stringFromData(namespace),
```
Comment (francisbouvier) [inline src/netsurf.zig:379]: I think we could try to shorter those functions as they are used a lot, like `strFromData` and `strToData`. It's not a lot but still 3 chars less and it's as clear as the full version I think.

Let's do that in a separate PR.

File: src/netsurf.zig [hunk 10]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+    if (n == null) {
+        return null;
+    }
+
+    // cast [*c]c.dom_node into *Attribute
+    return @as(*Attribute, @ptrCast(n));
+}
+
+// Convert a parser pointer to a public dom_node pointer.
+fn toDOMNode(comptime T: type, v: *T) [*c]c.dom_node {
```
Comment (francisbouvier) [inline src/netsurf.zig:392]: I think `toNodeInternal` is better to avoid ambiguity with `Node`?

File: src/netsurf.zig [hunk 11]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+    const err = c._dom_namednodemap_set_named_item(
+        nnm,
+        toDOMNode(Attribute, attr),
+        &n,
+    );
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]: Inliner

File: src/netsurf.zig [hunk 12]
```diff
@@ -336,6 +336,131 @@ fn DOMErr(except: DOMException) DOMError!void {
...
+    const err = c._dom_namednodemap_set_named_item_ns(
+        nnm,
+        toDOMNode(Attribute, attr),
+        &n,
+    );
+    try DOMErr(err);
+
+    if (n == null) {
+        return null;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]: Inliner

File: src/netsurf.zig (hunk 7 above)
Comment (krichprollsch) [reply src/netsurf.zig:350]: but in the contrary it's a public version of node, not an internal one...

File: src/netsurf.zig (hunk 9 above)
Comment (krichprollsch) [reply src/netsurf.zig:379]: https://github.com/Browsercore/browsercore/issues/108

File: src/netsurf.zig (hunk 7 above)
Comment (krichprollsch) [reply src/netsurf.zig:350]: after discussion, let use `NodeExternal`

Comment (krichprollsch) [reply src/netsurf.zig:350]: https://github.com/Browsercore/browsercore/issues/109

File: src/netsurf.zig (hunk 8 above)
Comment (krichprollsch) [reply src/netsurf.zig:356]: https://github.com/Browsercore/browsercore/issues/109

File: src/netsurf.zig (hunk 10 above)
Comment (krichprollsch) [reply src/netsurf.zig:392]: https://github.com/Browsercore/browsercore/issues/109#issuecomment-1840430816
